java-symbolic-execution/src/main/java/org/sonar/java/se/checks/DenialOfServiceXMLCheck.java

```diff
@@ -0,0 +1,120 @@
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+import static org.sonar.java.se.checks.XxeProcessingCheck.NEW_DOCUMENT_BUILDER;
+import static org.sonar.java.se.checks.XxeProcessingCheck.PARSING_METHODS;
+
+@Rule(key = "S6376")
+public class DenialOfServiceXMLCheck extends SECheck {
+
+  private static final MethodMatchers PARSING_METHODS_DOS = MethodMatchers.or(
+    PARSING_METHODS,
+    // When "newDocumentBuilder" is called on the factory, there is no way to secure the processing anymore
+    NEW_DOCUMENT_BUILDER
+  );
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    private final CheckerContext context;
+
+    private PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+      this.context = context;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree mit) {
+      if (PARSING_METHODS_DOS.matches(mit)) {
+        SymbolicValue peek = programState.peekValue(mit.arguments().size());
+
+        if (peek instanceof XxeSymbolicValue) {
+          XxeSymbolicValue xxeSymbolicValue = (XxeSymbolicValue) peek;
+          reportIfUnSecured(context, xxeSymbolicValue, programState.getConstraints(xxeSymbolicValue));
+        }
+      }
+    }
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    ProgramState endState = context.getState();
+    if (endState.exitingOnRuntimeException()) {
+      return;
+    }
+
+    // We want to report only when the unsecured factory is returned, if it is the case, it will be on the top of the stack.
+    SymbolicValue peek = endState.peekValue();
+    if (peek instanceof XxeSymbolicValue) {
+      XxeSymbolicValue xxeSV = (XxeSymbolicValue) peek;
+      reportIfUnSecured(context, xxeSV, endState.getConstraints(xxeSV));
+    }
+  }
+
+  private void reportIfUnSecured(CheckerContext context, XxeSymbolicValue xxeSV, @Nullable ConstraintsByDomain constraintsByDomain) {
+    if (!xxeSV.isField && isUnSecuredByProperty(constraintsByDomain)) {
+      context.reportIssue(getIssueLocation(context, xxeSV),
+        this,
+        "Enable XML parsing limitations to prevent Denial of Service attacks.");
+    }
+  }
+
+  private static Tree getIssueLocation(CheckerContext context, XxeSymbolicValue xxeSV) {
+    return FlowComputation.flowWithoutExceptions(context.getNode(), xxeSV, c -> c == FeatureSecureProcessing.UNSECURED,
+      Collections.singletonList(FeatureSecureProcessing.class), FlowComputation.FIRST_FLOW)
+      .stream()
+      .findFirst()
+      .flatMap(f -> f.elements().stream().findFirst())
```

Comment:
I think it's because we only need one tree to report. Having multiple path does not necessarily help the user to understand the issue, and we only have to find one to prove there is a problem. Anyway, from what we can see, we don't build flows in our case.

Reply:
**In one flow,** the logic in `FlowComputation.flowWithoutExceptions` only returns a location when the constraint changes. It means that anyway only one location will be returned for the code:
```
builder.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, false); // Issue here
builder.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, false); 
```
Another example, where it will return two locations (because the constraint changes two times):
```
    builder.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, false); // Second location in the flow
    builder.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
    builder.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, false); // Issue here, last time the constraint changed
```

For this rule, there is only one way to trigger an issue, I think it is fine to support **only one location per flow**, since the others cases are unlikely to happen.

Now, what about multiple flows? I realized that we are in fact possibly raising multiple issues, one each time we reach the end of execution/parsing method with an unsecured factory.
The SE already filters when we raise two issues at the same location. We still sometimes raise two issues, if we set the property to false two times in two different flows. 
I added more test cases, I guess we should discuss what to do at this point.